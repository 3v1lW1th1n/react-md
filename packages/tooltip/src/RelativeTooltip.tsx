import * as React from "react";
import * as PropTypes from "prop-types";
import cn from "classnames";

import {
  TooltipPosition,
  DEFAULT_ENTER_DURATION,
  DEFAULT_LEAVE_DURATION,
  DEFAULT_SHOW_DELAY,
  DEFAULT_POSITION,
} from "./constants";
import { default as Tooltip, ITooltipProps, ITooltipDefaultProps } from "./Tooltip";

export interface IRelativeTooltipProps extends ITooltipProps {
  /**
   * An optional delay before showing the tooltip when the user focuses or moves the mouse into the tooltip's container.
   */
  delay?: number;

  /**
   * The enter duration in milliseconds for the tooltip to fully animate into view. This should match whatever value is
   * set for `$rmd-tooltip-enter-duration`. A manual timeout is used instead of `onTransitionEnd` to handle cancel
   * animations easier.
   */
  enterDuration?: number;

  /**
   * The leave duration in milliseconds for the tooltip to fully animate into view. This should match whatever value is
   * set for `$rmd-tooltip-leave-duration`. A manual timeout is used instead of `onTransitionEnd` to handle cancel
   * animations easier.
   */
  leaveDuration?: number;
}

export interface IRelativeTooltipDefaultProps extends ITooltipDefaultProps {
  delay: number;
  enterDuration: number;
  leaveDuration: number;
}

export type RelativeTooltipWithDefaultProps = IRelativeTooltipProps & IRelativeTooltipDefaultProps;

export interface IRelativeTooltipState {
  keyboard: boolean;
  animatingIn: boolean;
  animatingOut: boolean;
  visible: boolean;
}

export default class RelativeTooltip extends React.Component<
  IRelativeTooltipProps,
  IRelativeTooltipState
> {
  public static propTypes = {
    id: PropTypes.string.isRequired,
    style: PropTypes.object,
    className: PropTypes.string,
    children: PropTypes.node.isRequired,
    delay: PropTypes.number,
    dense: PropTypes.bool,
    position: PropTypes.oneOf(["left", "top", "right", "bottom"]),
  };

  public static defaultProps: IRelativeTooltipDefaultProps = {
    delay: DEFAULT_SHOW_DELAY,
    enterDuration: DEFAULT_ENTER_DURATION,
    leaveDuration: DEFAULT_LEAVE_DURATION,
    dense: false,
    position: DEFAULT_POSITION,
    lineWrap: false,
  };

  /**
   * The tooltip's container element. This should be a focusable element so that keyboard users will be able
   * to view the tooltip when focusing the element.
   */
  private container: HTMLElement | null;

  /**
   * This is an animation frame that is used for the enter and leave transitions.
   */
  private frame?: number;

  /**
   * The current timeout for showing the tooltip if there is a delay.
   */
  private timeout?: number;
  private transitionTimeout?: number;

  constructor(props: RelativeTooltipWithDefaultProps) {
    super(props);

    this.container = null;
    this.state = {
      keyboard: false,
      animatingIn: false,
      animatingOut: false,
      visible: false,
    };
  }

  public componentDidMount() {
    this.init();
  }

  public componentWillUnmount() {
    this.deinit();
  }

  public render() {
    const { visible, animatingIn, animatingOut } = this.state;
    const { children, className, enterDuration, leaveDuration, delay, ...props } = this
      .props as RelativeTooltipWithDefaultProps;

    return (
      <Tooltip
        {...props}
        className={cn(
          {
            "rmd-tooltip--active": visible || animatingIn || animatingOut,
            "rmd-tooltip--enter": animatingIn,
            "rmd-tooltip--leave": animatingOut,
          },
          className
        )}
        visible={visible}
      >
        {children}
      </Tooltip>
    );
  }

  /**
   * Initializes the tooltip by attempting to find its container element and then attaching the event listeners
   * for showing and hiding the tooltip.
   */
  private init = () => {
    const { id } = this.props;
    const container = document.querySelector(`[aria-describedby="${id}"]`) as HTMLElement;
    if (!container) {
      if (process.env.NODE_ENV !== "production") {
        throw new Error(
          'A tooltip\'s container must have the attribute `aria-describedby="TOOLTIP_ID"` for accessibility ' +
            `but none were found for a tooltip with id: \`${id}\``
        );
      }

      this.container = null;
      return;
    }

    container.addEventListener("mouseenter", this.handleMouseEnter);
    container.addEventListener("mouseleave", this.handleMouseLeave);
    container.addEventListener("keydown", this.handleKeyDown);
    container.addEventListener("keyup", this.handleKeyUp);
    container.addEventListener("blur", this.handleBlur);
    this.container = container;
  };

  /**
   * This is normally just called when the component unmounts. It will remove all the event listeners
   * from the container element if it exists.
   */
  private deinit = () => {
    if (!this.container) {
      return;
    }

    this.clear();
    this.container.removeEventListener("mouseenter", this.handleMouseEnter);
    this.container.removeEventListener("mouseleave", this.handleMouseLeave);
    this.container.removeEventListener("keydown", this.handleKeyDown);
    this.container.removeEventListener("keyup", this.handleKeyUp);
    this.container.removeEventListener("blur", this.handleBlur);
  };

  private clear = () => {
    if (this.timeout) {
      window.clearTimeout(this.timeout);
      this.timeout = undefined;
    }

    if (this.frame) {
      window.cancelAnimationFrame(this.frame);
      this.frame = undefined;
    }

    if (this.transitionTimeout) {
      window.clearTimeout(this.transitionTimeout);
      this.transitionTimeout = undefined;
    }
  };

  /**
   * This will conditionally show the tooltip from one of the container's event listeners. If there is a delay, it
   * will create a timeout for the delay duration and then start showing the tooltip. If there is no delay, the tooltip
   * will immediately be shown.
   */
  private show = (keyboard: boolean) => {
    const { delay } = this.props as RelativeTooltipWithDefaultProps;
    const { visible } = this.state;
    if (this.timeout || !this.container || (visible && this.state.keyboard !== keyboard)) {
      return;
    }

    if (process.env.NODE_ENV !== "production") {
      const tooltip = document.getElementById(this.props.id);

      // when in dev mode, make sure that the container element has a relative position so that the tooltip
      // will appear "fixed" to the container element. This will have to be updated when I implement
      // tooltip portal support
      const { position } = window.getComputedStyle(this.container);
      if (tooltip && this.container.contains(tooltip) && position !== "relative") {
        // tslint:disable-next-line:no-console
        console.error(
          "A tooltip's container must have `position: relative` as a style for a tooltip to appear but its position " +
            `is: \`${position}\`. An inline style has been applied, but your container must be updated before going ` +
            "into production as this functionality will be removed. Tooltip Container: ",
          this.container
        );

        this.container.style.position = "relative";
      }
    }

    this.timeout = window.setTimeout(() => {
      this.timeout = undefined;
      this.animateIn();
    }, Math.max(0, delay));

    if (this.state.keyboard !== keyboard || this.state.animatingOut || this.state.animatingIn) {
      this.setState({ keyboard, animatingOut: false, animatingIn: false });
    }
  };

  private animateIn = () => {
    const { enterDuration } = this.props as RelativeTooltipWithDefaultProps;
    this.setState({ animatingIn: true });
    this.frame = window.requestAnimationFrame(() => {
      this.frame = undefined;
      this.setState({ visible: true });

      this.transitionTimeout = window.setTimeout(() => {
        this.transitionTimeout = undefined;
        this.setState({ animatingIn: false });
      }, Math.max(0, enterDuration));
    });
  };

  private hide = (keyboard: boolean) => {
    const { visible } = this.state;
    const { leaveDuration } = this.props as RelativeTooltipWithDefaultProps;
    this.clear();
    if (visible && this.state.keyboard === keyboard) {
      this.transitionTimeout = window.setTimeout(() => {
        this.transitionTimeout = undefined;
        this.setState({ animatingOut: false });
      }, Math.max(0, leaveDuration));
      this.setState({
        animatingOut: true,
        animatingIn: false,
        keyboard: false,
        visible: false,
      });
    } else if (this.state.animatingIn) {
      this.setState({ animatingIn: false });
    }
  };

  private handleMouseEnter = () => {
    this.show(false);
  };

  private handleMouseLeave = () => {
    this.hide(false);
  };

  /**
   * This is way to capture focus events on the tooltip's container element. When the container is
   * focused by keyboard navigation, the tooltip should be shown.
   */
  private handleKeyUp = (e: KeyboardEvent) => {
    if (e.key === "Tab") {
      this.show(true);
    }
  };

  private handleKeyDown = (e: KeyboardEvent) => {
    if (e.key === "Escape") {
      // always hide the tooltip immediately if escape is pressed
      this.hide(this.state.keyboard);
    }
  };

  private handleBlur = () => {
    this.hide(true);
  };
}
