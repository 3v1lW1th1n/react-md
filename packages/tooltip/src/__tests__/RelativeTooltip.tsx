/* tslint:disable:max-line-length */
import * as React from "react";
import { mount } from "enzyme";

import RelativeTooltip from "../RelativeTooltip";

describe("RelativeTooltip", () => {
  const id = "tooltip-id";
  const addEventListener = jest.fn();
  const removeEventListener = jest.fn();
  beforeEach(() => {
    jest.useFakeTimers();
    addEventListener.mockClear();
    removeEventListener.mockClear();
    jest.spyOn(document, "querySelector").mockImplementation(query => {
      if (query === `[aria-describedby="${id}"]`) {
        return { addEventListener, removeEventListener };
      }

      return null;
    });
  });

  it("should attach/remove the mouseenter, mouseleave, keydown, keyup, and blur events to the container element on mount/unmount", () => {
    const tooltip = mount(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);

    expect(addEventListener.mock.calls.length).toBe(5);

    tooltip.unmount();
    expect(removeEventListener.mock.calls.length).toBe(5);
  });

  it("should throw an error if there are no element found with the aria-describedby query selector", () => {
    jest.spyOn(console, "error").mockImplementation(() => false);
    let caughtInfo: React.ErrorInfo;
    let caughtError: Error;

    class ErrorCatch extends React.Component {
      public state = { error: false };
      public componentDidCatch(error: Error, info: React.ErrorInfo) {
        caughtInfo = info;
        caughtError = error;
        this.setState({ error: true });
      }

      public render() {
        if (this.state.error) {
          return null;
        }

        return <RelativeTooltip id="fake-id">Tooltip</RelativeTooltip>;
      }
    }

    mount(<ErrorCatch />);
    expect(caughtError).toEqual(
      new Error(
        'A tooltip\'s container must have the attribute `aria-describedby="TOOLTIP_ID"` for accessibility ' +
          "but none were found for a tooltip with id: `fake-id`"
      )
    );
    expect(caughtInfo).toBeDefined();
  });

  describe("entering", () => {
    it("should do the enter animation correctly from the component show function", () => {
      jest.spyOn(window, "getComputedStyle").mockImplementation(() => ({ position: "relative" }));
      jest
        .spyOn(window, "requestAnimationFrame")
        .mockImplementation(cb => window.setTimeout(cb, 0));

      const tooltip = mount<RelativeTooltip>(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);
      expect(tooltip.state("visible")).toBe(false);
      expect(tooltip.state("animatingIn")).toBe(false);
      expect(tooltip.state("animatingOut")).toBe(false);
      expect(tooltip.state("keyboard")).toBe(false);

      tooltip.instance().show(false);
      expect(window.setTimeout.mock.calls.length).toBe(1);
      expect(window.requestAnimationFrame.mock.calls.length).toBe(0);
      expect(tooltip.state("visible")).toBe(false);
      expect(tooltip.state("animatingIn")).toBe(false);
      expect(tooltip.state("animatingOut")).toBe(false);
      expect(tooltip.state("keyboard")).toBe(false);

      jest.runOnlyPendingTimers();
      expect(window.requestAnimationFrame.mock.calls.length).toBe(1);
      expect(window.setTimeout.mock.calls.length).toBe(2);
      expect(tooltip.state("visible")).toBe(false);
      expect(tooltip.state("animatingIn")).toBe(true);
      expect(tooltip.state("animatingOut")).toBe(false);
      expect(tooltip.state("keyboard")).toBe(false);

      jest.runOnlyPendingTimers();
      expect(window.requestAnimationFrame.mock.calls.length).toBe(1);
      expect(window.setTimeout.mock.calls.length).toBe(3);
      expect(tooltip.state("visible")).toBe(true);
      expect(tooltip.state("animatingIn")).toBe(true);
      expect(tooltip.state("animatingOut")).toBe(false);
      expect(tooltip.state("keyboard")).toBe(false);

      jest.runOnlyPendingTimers();
      expect(window.requestAnimationFrame.mock.calls.length).toBe(1);
      expect(window.setTimeout.mock.calls.length).toBe(3);
      expect(tooltip.state("visible")).toBe(true);
      expect(tooltip.state("animatingIn")).toBe(false);
      expect(tooltip.state("animatingOut")).toBe(false);
      expect(tooltip.state("keyboard")).toBe(false);
    });

    it("should not restart the enter timeout if the show function is called again before the initial delay finishes", () => {
      jest.spyOn(window, "getComputedStyle").mockImplementation(() => ({ position: "relative" }));

      const tooltip = mount<RelativeTooltip>(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);

      tooltip.instance().show(false);
      expect(window.setTimeout.mock.calls.length).toBe(1);
      expect(window.clearTimeout.mock.calls.length).toBe(0);

      tooltip.instance().show(false);
      expect(window.setTimeout.mock.calls.length).toBe(1);
      expect(window.clearTimeout.mock.calls.length).toBe(0);
    });

    it("should not restart the enter timeout if the show function is called again while already visible and the keyboard boolean is different from the state value", () => {
      jest.spyOn(window, "getComputedStyle").mockImplementation(() => ({ position: "relative" }));

      const tooltip = mount<RelativeTooltip>(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);

      tooltip.instance().show(false);
      expect(tooltip.state("keyboard")).toBe(false);
      expect(window.setTimeout.mock.calls.length).toBe(1);
      expect(window.clearTimeout.mock.calls.length).toBe(0);
      jest.runAllTimers();

      expect(window.setTimeout.mock.calls.length).toBe(5);

      tooltip.instance().show(true);
      expect(window.setTimeout.mock.calls.length).toBe(5);
      expect(window.clearTimeout.mock.calls.length).toBe(0);
    });

    it("should set the animatingIn or animatingOut state to false if they are true when the show animation starts", () => {
      jest.spyOn(window, "getComputedStyle").mockImplementation(() => ({ position: "relative" }));

      const tooltip = mount<RelativeTooltip>(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);
      tooltip.setState({ animatingIn: true });

      tooltip.instance().show(false);
      expect(tooltip.state("animatingIn")).toBe(false);

      jest.runAllTimers();
      tooltip.setState({ visible: false, animatingOut: true });

      tooltip.instance().show(false);
      expect(tooltip.state("animatingOut")).toBe(false);
      jest.runAllTimers();
    });

    it("should update the keyboard state if it was different than the state keyboard", () => {
      jest.spyOn(window, "getComputedStyle").mockImplementation(() => ({ position: "relative" }));
      const tooltip = mount<RelativeTooltip>(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);
      expect(tooltip.state("keyboard")).toBe(false);

      tooltip.instance().show(true);
      jest.runAllTimers();
      expect(tooltip.state("keyboard")).toBe(true);
      expect(tooltip.state("visible")).toBe(true);

      tooltip.setState({ visible: false });
      tooltip.instance().show(false);
      jest.runAllTimers();
      expect(tooltip.state("keyboard")).toBe(false);
    });
  });

  describe("leaving", () => {
    it("should clear all timeouts and animation frames immediately", () => {
      jest.spyOn(window, "cancelAnimationFrame").mockImplementation(t => window.clearTimeout(t));
      const tooltip = mount<RelativeTooltip>(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);
      tooltip.instance().timeout = 1;
      tooltip.instance().transitionTimeout = 2;
      tooltip.instance().frame = 3;

      tooltip.instance().hide(false);
      expect(window.clearTimeout.mock.calls.length).toBe(3);
      expect(window.clearTimeout).toBeCalledWith(1);
      expect(window.clearTimeout).toBeCalledWith(2);
      expect(window.clearTimeout).toBeCalledWith(3);

      expect(tooltip.instance().timeout).toBeUndefined();
      expect(tooltip.instance().transitionTimeout).toBeUndefined();
      expect(tooltip.instance().fame).toBeUndefined();
    });

    it("should start the transitionTimeout if visible and the keyboard bool is equal the the state value", () => {
      const tooltip = mount(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);
      tooltip.setState({ keyboard: false, visible: true });

      tooltip.instance().hide(false);
      expect(setTimeout.mock.calls.length).toBe(1);
      expect(tooltip.state("visible")).toBe(false);
      expect(tooltip.state("animatingIn")).toBe(false);
      expect(tooltip.state("animatingOut")).toBe(true);
      expect(tooltip.state("keyboard")).toBe(false);

      jest.runOnlyPendingTimers();
      expect(setTimeout.mock.calls.length).toBe(1);
      expect(tooltip.state("visible")).toBe(false);
      expect(tooltip.state("animatingIn")).toBe(false);
      expect(tooltip.state("animatingOut")).toBe(false);
      expect(tooltip.state("keyboard")).toBe(false);
    });

    it("should not start the leave transition if the keyboard state is different", () => {
      const tooltip = mount(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);
      tooltip.setState({ keyboard: true, visible: true });

      tooltip.instance().hide(false);
      expect(setTimeout.mock.calls.length).toBe(0);

      tooltip.setState({ keyboard: false });
      tooltip.instance().hide(true);
      expect(setTimeout.mock.calls.length).toBe(0);
    });
  });

  it("should render correctly at the different stages of animation", () => {
    jest.spyOn(window, "requestAnimationFrame").mockImplementation(cb => window.setTimeout(cb, 0));
    const tooltip = mount(<RelativeTooltip id={id}>Tooltip</RelativeTooltip>);
    expect(tooltip.render()).toMatchSnapshot();

    tooltip.instance().show(true);
    expect(tooltip.render()).toMatchSnapshot();
    jest.runOnlyPendingTimers();
    expect(tooltip.render()).toMatchSnapshot();
    jest.runOnlyPendingTimers();
    expect(tooltip.render()).toMatchSnapshot();
    jest.runAllTimers();
    expect(tooltip.render()).toMatchSnapshot();

    tooltip.instance().hide(true);
    expect(tooltip.render()).toMatchSnapshot();
    jest.runAllTimers();
    expect(tooltip.render()).toMatchSnapshot();
  });
});
