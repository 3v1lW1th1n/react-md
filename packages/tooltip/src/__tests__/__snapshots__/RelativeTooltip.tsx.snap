// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelativeTooltip should render correctly at the different stages of animation 1`] = `
<span
  class="rmd-tooltip rmd-tooltip--bottom"
  id="tooltip-id"
  role="tooltip"
>
  Tooltip
</span>
`;

exports[`RelativeTooltip should render correctly at the different stages of animation 2`] = `
<span
  class="rmd-tooltip rmd-tooltip--bottom"
  id="tooltip-id"
  role="tooltip"
>
  Tooltip
</span>
`;

exports[`RelativeTooltip should render correctly at the different stages of animation 3`] = `
<span
  class="rmd-tooltip rmd-tooltip--bottom rmd-tooltip--active rmd-tooltip--enter"
  id="tooltip-id"
  role="tooltip"
>
  Tooltip
</span>
`;

exports[`RelativeTooltip should render correctly at the different stages of animation 4`] = `
<span
  class="rmd-tooltip rmd-tooltip--visible rmd-tooltip--bottom rmd-tooltip--active rmd-tooltip--enter"
  id="tooltip-id"
  role="tooltip"
>
  Tooltip
</span>
`;

exports[`RelativeTooltip should render correctly at the different stages of animation 5`] = `
<span
  class="rmd-tooltip rmd-tooltip--visible rmd-tooltip--bottom rmd-tooltip--active"
  id="tooltip-id"
  role="tooltip"
>
  Tooltip
</span>
`;

exports[`RelativeTooltip should render correctly at the different stages of animation 6`] = `
<span
  class="rmd-tooltip rmd-tooltip--bottom rmd-tooltip--active rmd-tooltip--leave"
  id="tooltip-id"
  role="tooltip"
>
  Tooltip
</span>
`;

exports[`RelativeTooltip should render correctly at the different stages of animation 7`] = `
<span
  class="rmd-tooltip rmd-tooltip--bottom"
  id="tooltip-id"
  role="tooltip"
>
  Tooltip
</span>
`;
